metaflow/plugins/aws/awslambda/lambda_runner.py

```diff
@@ -0,0 +1,314 @@
+        super(LogTailer, self).__init__()
+        self.daemon = True
+
+    def stop_tail(self, timeout_seconds, log_end_token):
+        """
+        Stop tailing.
+
+        After Lambda invocation is done, there is generally no way for us to know if
+        there are any log messages that were written by Lambda to Cloudwatch, but
+        haven't yet been read by this LogTailer.
+
+        So we use the following strategy to decide when to stop reading: Lambda will
+        write a unique log_end_token at the end. On the reading end, after lambda
+        finishes, we'll read until we encounter that token, or until timeout_second
+        expires. The timeout is useful for the case where Lambda dies unexpectedly and
+        doesn't get a chance to write the token.
+        """
+        # Slightly awkward naming here to make sure there is no conflict with Thread
+        # internal members.
+        self._stop_deadline = time.time() + timeout_seconds
+        self._log_end_token = log_end_token
+
+    def run(self):
+        logs = BatchLogs(self._log_group, self._log_stream, sleep_on_no_data=1)
+        seen_start_token = False
+        while True:
+            for line in logs:
+                if line:
+                    if self._log_end_token is not None and line == self._log_end_token:
+                        return
+                    else:
+                        if (
+                            not seen_start_token
+                            and line.strip() == self._log_start_token
+                        ):
+                            seen_start_token = True
+                        elif seen_start_token:
+                            self._echo(line)
+                else:
+                    if self._stop_deadline and self._stop_deadline < time.time():
+                        return
+
+
+def lambda_name(flow_name, step_name):
+    """ Construct lambda name from flow and step name """
+    return "-".join(["mflambda", flow_name, step_name])
+
+
+def ensure_lambda(
+    client,
+    flow_name,
+    step_name,
+    timeout_seconds: int,
+    memory_mb: Optional[int],
+    cpu: Optional[float],
+    role_arn,
+    image_uri,
+    force_fetch_image: bool,
+    echo: Callable[[str], None],
+) -> str:
+    """
+    Make sure Lambda for the step exists. If it does, but resources or timeout
+    configuration is not what we expect, update the lambda configuration.
+    """
+
+    # Compute lambda name
+    name = lambda_name(flow_name, step_name)
+
+    # Compute Lambda memory_size.
+    #
+    # For Lambda, you can't specify CPU and mem independently, you always get
+    # 1769 (MB/vCPU) [1]. In @resources deco, we allow users to specify cpu or mem or
+    # both. So when it comes to lambda, we'll set memory_size to a value that would
+    # satisfy the max of both requirements.
+    #
+    # [1] https://docs.aws.amazon.com/lambda/latest/dg/configuration-memory.html
+    if cpu is not None and memory_mb is not None:
```

Comment:
This calculation runs the risk of becoming stale as AWS introduces new resource instances for Lambda. Is there a need for us to do this validation? Given that launching an AWS Lambda function is quick and easy - we can simply launch the function with the user-defined resources and return a properly formatted error if the values are incorrect.

Reply:
Re: validation, see the other comment. 

As for this math, lambda doesn't allow you to specify vCPUs directly, only memory, and you get X cpu for Y memory. I figured it would be nice if we still allowed users to specify CPU using `@resources` and did the math for them. The alternative is that we could say that `@resources` deco only supports specifying memory size for lambda, and we can remove this math. But users would have to do it anyway, and it is not trivial to dig it out of AWS docs.

Reply:
I think we can simply allow our users to only specify `memory`. As AWS changes the math on its end, it _might_ be tricky for us to guarantee backward compatibility.

Reply:
removed the cpu->memory_size math in c799b10